classes/security/authorization/internal/ApiTokenDecodingMiddleware.inc.php

```diff
@@ -36,27 +38,58 @@ public function __construct(APIHandler $handler) {
 	 */
 	protected function _decode($slimRequest) {
 		$secret = Config::getVar('security', 'api_key_secret', '');
+
+		if ($secret === '') {
+			$request = $this->_handler->getRequest();
+			return $request->getRouter()
+				->handleAuthorizationFailure(
+					$request,
+					'api.api_key_secret.should.be.filled'
+				);
+		}
+
 		if ($secret !== '' && !is_null($jwt = $slimRequest->getQueryParam('apiToken'))) {
 			try {
-				$apiToken = JWT::decode($jwt, $secret, array('HS256'));
+				$apiToken = JWT::decode($jwt, $secret, ['HS256']);
 				$this->_handler->setApiToken($apiToken);
 				return true;
 			} catch (Exception $e) {
-				// If JWT decoding fails, it throws an
-				// 'UnexpectedValueException'.  If JSON decoding fails
-				// (of the JWT payload), it throws a 'DomainException'.
-				if (is_a($e, 'UnexpectedValueException') || is_a($e, 'DomainException')) {
+				/**
+				 * If JWT decoding fails, it throws an 'UnexpectedValueException'.
+				 * If JSON decoding fails (of the JWT payload), it throws a 'DomainException'.
+				 * If token couldn't verified, it throws a 'SignatureInvalidException'.
+				 */
+				if (is_a($e, SignatureInvalidException::class)) {
+					$request = $this->_handler->getRequest();
+					return $request->getRouter()
+						->handleAuthorizationFailure(
+							$request,
+							'api.invalid_token_signature'
+						);
+				}
+
+				if (is_a($e, 'UnexpectedValueException') ||
+					is_a($e, 'DomainException')
+				) {
 					$request = $this->_handler->getRequest();
-					$router = $request->getRouter();
-					$result = $router->handleAuthorizationFailure($request, $e->getMessage());
-					switch(1) {
-						case is_string($result): return $result;
-						case is_a($result, 'JSONMessage'): return $result->getString();
-						default:
-							assert(false);
-							return null;
+					$result = $request->getRouter()
+						->handleAuthorizationFailure(
+							$request,
+							$e->getMessage()
+						);
+
+					if (is_string($result)) {
+						return $result;
+					}
+
+					if (is_a($result, 'JSONMessage')) {
+						return $result->getString();
 					}
+
+					assert(false);
+					return null;
```

Comment:
I think this is all some leftover logic from older handlers that can be removed. Also, let's use a defined locale key and put the `$e->message()` into that. Right now, the response to something like this comes out like this:

```
{
    "error": "Malformed UTF-8 characters",
    "errorMessage": "##Malformed UTF-8 characters##"
}
```

We want the `error` property in the response object to be a pre-defined error code, no matter what language the server is running. So let's do something like this. First, create a locale key that will act as a pre-defined error code, but allow us to pass the exception message into it:

```
msgid "api.403.unexpectedToken"
msgstr "The apiToken could not be decoded because of the following error: {$error}"
```

Then pass something like this in the return result:

```php
$request->getRouter()->handleAuthorizationFailure($request, 'api.403.unexpectedToken', ['error' => $e->getMessage()]);
```

In order for this to work, you will need to extend `APIRouter::handleAuthorizationFailure()` to accept an optional third parameter, which will be passed to the `__()` translation function. (This will require you to extend the base class method too.)

Reply:
> I think this is all some leftover logic from older handlers that can be removed. 

Sorry, I wasn't clear about this. What i mean is that the logic checking for `is_string` and whether something is a `JSONMessage` is old and can be removed. In fact, when `APIRouter::handleAuthorizationFailure()` is called, the program [exits](https://github.com/pkp/pkp-lib/blob/master/classes/core/APIRouter.inc.php#L150) so there's no need for any further handling.